common/migration/read.go

```diff
@@ -0,0 +1,163 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package migration
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+)
+
+const (
+	ReaderRolloutCallOnlyOldAndReturnOld ReaderRolloutState = "only-call-old"
+	ReaderRolloutCallBothAndReturnOld    ReaderRolloutState = "call-both-return-old"
+	ReaderRolloutCallBothAndReturnNew    ReaderRolloutState = "call-both-return-new"
+	ReaderRolloutCallOnlyNewAndReturnNew ReaderRolloutState = "only-call-new"
+)
+
+type resultPair[T any] struct {
+	res T
+	err error
+}
+
+func (c *readerImpl[T]) ReadAndReturnActive(
+	ctx context.Context,
+	constraints Constraints,
+	old func(ctx context.Context) (*T, error),
+	new func(ctx context.Context) (*T, error),
```

Comment:
from personal fiddling with types:  `*T` is I think almost always wrong / more confusing than it's worth.  just use `T`.

I get the desire to make things nullable on error, but if they pass a nullable-if-err type like `*myresult`, they now have a `**myresult` response type, and `!= nil` is no longer "data exists" - they may need to `!= nil { *res != nil { ...`